File: 1706-min-cost-to-connect-all-points/min-cost-to-connect-all-points.java
class Solution {
    public int minCostConnectPoints(int[][] points) {
        HashMap<Integer, List<List<Integer>>> mapper = new HashMap<Integer, List<List<Integer>>>();
        for (int i = 0; i < points.length; i++){
            List<Integer> a = Arrays.asList(points[i][0], points[i][1]);
            for (int j = i; j < points.length; j++){
                List<Integer> b = Arrays.asList(points[j][0], points[j][1]);
                List<Integer> temp = Arrays.asList(Math.abs(a.get(0) - b.get(0)) + Math.abs(a.get(1) - b.get(1)), j);
                List<List<Integer>> t = mapper.getOrDefault(i, new ArrayList<List<Integer>>());
                t.add(temp);
                mapper.put(i, t); 
                temp = Arrays.asList(Math.abs(a.get(0) - b.get(0)) + Math.abs(a.get(1) - b.get(1)), i);
                t = mapper.getOrDefault(j, new ArrayList<List<Integer>>());
                t.add(temp);
                mapper.put(j, t);
            }
        }

        PriorityQueue<Pair<Integer,Integer>> q = new PriorityQueue<Pair<Integer,Integer>>(Comparator.comparingInt(a->a.getKey()));
        q.add(new Pair<Integer,Integer>(0,0));
        HashSet<Integer> visited = new HashSet<Integer>();
        int result = 0;
        while (visited.size() < points.length){
            Pair<Integer,Integer> temp = q.poll();
            if(visited.contains(temp.getValue())) continue;
            result += temp.getKey();
            visited.add(temp.getValue());
            for (List<Integer> t: mapper.get(temp.getValue()))
                if (!visited.contains(t.get(1))) q.add(new Pair<Integer,Integer>(t.get(0),t.get(1)));
        }
        return result;
    }
}

File: 1706-min-cost-to-connect-all-points/min-cost-to-connect-all-points.cpp
class Solution {
public:
    struct comparator{
        bool operator()(const pair<int, int>& t1, const pair<int, int>& t2){
            return t1.first >= t2.first;
        }
    };


    int minCostConnectPoints(vector<vector<int>>& points) {
        unordered_map<int, vector<vector<int>>> mapper;
        for (int i = 0; i < points.size(); i++){
            vector<int> a = points[i];
            for (int j = i; j < points.size(); j++){
                vector<int> b = points[j];
                vector<int> temp = {abs(a[0] - b[0]) + abs(a[1] - b[1]), j};
                mapper[i].push_back(temp); 
                temp = {abs(a[0] - b[0]) + abs(a[1] - b[1]), i};
                mapper[j].push_back(temp);
            }
        }

        priority_queue<pair<int, int>, vector<pair<int, int>>, comparator> q;
        q.push(pair<int, int>(0,0));
        unordered_set<int> visited;
        int result = 0;
        while (visited.size() < points.size()){
            pair<int, int> temp = q.top();
            q.pop();
            if(visited.contains(temp.second)) continue;
            result += temp.first;
            visited.insert(temp.second);
            for (vector<int> t: mapper[temp.second])
                if (!visited.contains(t[1])) q.push(pair<int, int>(t[0],t[1]));
        }
        return result;
    }
};